src/tests/test_bootstrap_in_place.py

```diff
@@ -237,11 +255,150 @@ def test_bootstrap_in_place(
+
+            if not f"{CLUSTER_PREFIX}-worker-0" in node_readiness_map:
+                return False
+
+            return all(node_status for node_status in node_readiness_map.values())
+
+        except Exception as e:
+            print("got exception while validating operators: %s", e)
+            return False
+
+    @JunitTestCase()
+    def waiting_for_added_worker(self, controller: NodeController):
+        vm_ip = controller.master_ips[0][0]
+        try:
+            log.info("Waiting for worker to be added...")
+            waiting.wait(
+                self.worker_ready, sleep_seconds=20, timeout_seconds=60 * 60, waiting_for="worker node to be ready"
+            )
+            log.info("Day 2 worker addition finished successfully!")
+        finally:
+            self.log_collection(vm_ip)
+
+    def approve_csrs(self, kubeconfig, done: threading.Event):
+        log.info("Started background worker to approve CSRs when they appear...")
+        while not done.is_set():
+            for csr_name in get_unapproved_csr_names(kubeconfig):
+                try:
+                    log.info(f"Found unapproved CSR {csr_name}, approving...")
+                    approve_csr(kubeconfig, csr_name)
+                except:
+                    log.warning("Failed attempt to approve CSR, retrying...")
+                    continue
+
+            time.sleep(10)
+
+    @JunitTestSuite()
+    def test_bip_add_worker(
+        self,
+        controller: TerraformController,
+        controller_configuration: BaseNodeConfig,
+        cluster_configuration: ClusterConfig,
+    ):
+        image_path = self.prepare_installation(controller_configuration, cluster_configuration)
+        cluster_configuration.iso_download_path = image_path
+
+        with open(os.path.join(IBIP_DIR, "worker-live-iso.ign"), "w") as f:
+            f.write(self.render_worker_live_iso_ignition(INSTALLATION_DISK))
+
+        worker_image_path = self.embed("installer-image.iso", "worker-live-iso.ign", EMBED_IMAGE_NAME_WORKER)
+        cluster_configuration.worker_iso_download_path = worker_image_path
+
+        controller._create_nodes()
+
+        log.info("Starting master node...")
+        controller.start_node(node_name=f"{CLUSTER_PREFIX}-master-0")
+
+        self.waiting_for_installation_completion(controller, cluster_configuration, skip_logs=True)
+
+        image_path = self.embed("installer-image.iso", "worker-live-iso.ign", EMBED_IMAGE_NAME_WORKER)
+        controller.start_node(node_name=f"{CLUSTER_PREFIX}-worker-0")
+
+        # Start a background worker to approve CSRs
+        approve_csr_worker_done = threading.Event()
+        approve_csr_worker = threading.Thread(
+            target=self.approve_csrs,
+            args=(KUBE_CONFIG, approve_csr_worker_done),
+            # Don't hang if this thread is still running for some reason
+            daemon=True,
+        )
+        approve_csr_worker.start()
+
+        self.waiting_for_added_worker(controller)
```

Comment:
if this throws an exception you will not stop the approve_csr_worker thread,
consider adding try & finally